File: MDShockBannerViewOC/MDShockBannerViewOCDemo/MDShockBanner/MDShockBannerView.m
//
//  MDShockBannerView.m
//  MDShockBannerViewOCDemo
//
//  Created by Alan on 2018/4/27.
//  Copyright © 2018年 MD. All rights reserved.
//



#import <UIKit/UIKit.h>
#import "MDShockBannerView.h"
#import "UIView+MD.h"
#import "MDMaskView.h"
#import "MDBannerModel.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import "MDBannerCollectionViewCell.h"

#define mainW [[UIScreen mainScreen] bounds].size.width
#define mainH [[UIScreen mainScreen] bounds].size.height
#define cellID @"MDBannerCollectionViewCell"


@interface MDShockBannerView()<UICollectionViewDelegate,UICollectionViewDataSource>

@property(nonatomic,assign)CGFloat naxH;

@property(nonatomic, strong) UIImageView *topImage;
@property(nonatomic, strong) UIImageView *bottomImage;

@property(nonatomic, strong) UICollectionViewFlowLayout *flowLayout;

@property(nonatomic, strong) UICollectionView *banner;

@property(nonatomic, strong) MDMaskView *leftView;
@property(nonatomic, strong) MDMaskView *rightView;

@property(nonatomic, strong) UIView *bannerMaskView;

@property(nonatomic, strong) UIPageControl *pageControl;



@property(nonatomic, strong) NSTimer *timers;

@property(nonatomic, assign) CGFloat lastContentOffset;
@property(nonatomic, assign) NSInteger totalItemCount;
@property(nonatomic, assign) NSInteger draggingIndex;
@property(nonatomic, assign) NSInteger autoScrollIndex;

@end

@implementation MDShockBannerView


- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initShockBannerView];
    }
    return self;
}


-(void)initShockBannerView{
    self.autoScroll = YES;
    self.infiniteLoop = YES;
    self.scrollInterval = 3;
    
    [self addSubview:self.bottomImage];
    [self addSubview:self.topImage];
    [self addSubview:self.banner];
    [self addSubview:self.pageControl];

    self.topImage.maskView = self.bannerMaskView;
    
    [self.bannerMaskView addSubview:self.rightView];
    [self.bannerMaskView addSubview:self.leftView];
    
    [self.banner registerNib:[UINib nibWithNibName:cellID bundle:nil] forCellWithReuseIdentifier:cellID];
}







-(void)startTimer{
    [self invalidateTimer];
    self.autoScrollIndex = [self currentIndex];
    _timers = [NSTimer timerWithTimeInterval:_scrollInterval target:self selector:@selector(automaticScroll) userInfo:nil repeats:YES];
    [[NSRunLoop mainRunLoop] addTimer:_timers forMode:NSRunLoopCommonModes];
}


-(void)invalidateTimer{
    if (_timers != nil){
        [_timers invalidate];
        _timers = nil;
    }
}



-(void)automaticScroll{
    if (_totalItemCount == 0){return;}
    NSInteger currentIndex = [self currentIndex];
    NSInteger targetIndex = currentIndex + 1;
    UICollectionViewCell *cell = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:currentIndex inSection:0]];
    [UIView animateWithDuration:0.2 animations:^{
        cell.transform = CGAffineTransformMakeScale(0.85, 0.85);
    } completion:^(BOOL finished) {
        [self scrollToIndex:targetIndex];
    }];
 
}

-(void)scrollToIndex:(NSInteger)targetIndex{
    if (targetIndex >= _totalItemCount){
        [self.banner scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:_totalItemCount/2 inSection:0] atScrollPosition:UICollectionViewScrollPositionNone animated:NO];
        return;
    }
    [self.banner scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:targetIndex inSection:0] atScrollPosition:UICollectionViewScrollPositionNone animated:YES];
}


-(NSInteger)currentIndex{
    if (_banner.width == 0 || _banner.height == 0){return 0;}
    NSInteger index;

    index = (_banner.contentOffset.x + _flowLayout.itemSize.width/2)/_flowLayout.itemSize.width;
    return MAX(0, index);
}

-(NSInteger)pageControlIndexWithCurrentCellIndex:(NSInteger)index{
    return index % self.banners.count;
}



#pragma UICollectionViewDelegate,UICollectionViewDataSource

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return _totalItemCount;
}

-(UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    MDBannerCollectionViewCell * cell = [collectionView dequeueReusableCellWithReuseIdentifier:cellID forIndexPath:indexPath];
    NSInteger index = [self pageControlIndexWithCurrentCellIndex:indexPath.item];
    MDBannerModel *model = self.banners[index];
    [cell.i_img sd_setImageWithURL:[NSURL URLWithString:model.img]];
    return cell;
}


-(void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    if (self.delegate != nil && [self.delegate respondsToSelector:@selector(clickBanner:)]){
        [self.delegate clickBanner:[self pageControlIndexWithCurrentCellIndex:indexPath.item]];
    }
}



-(void)scrollViewDidScroll:(UIScrollView *)scrollView{
    CGFloat curretContentOffset = scrollView.contentOffset.x;
    self.pageControl.currentPage = [self pageControlIndexWithCurrentCellIndex:[self currentIndex]];
    if (scrollView.isDragging || scrollView.isDecelerating){
        NSInteger itemIndex = [self pageControlIndexWithCurrentCellIndex:_draggingIndex];
        if (_lastContentOffset > curretContentOffset){
            UICollectionViewCell *cell1 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:_draggingIndex inSection:0]];
            UICollectionViewCell *cell2 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_draggingIndex - 1 < 0) ? _totalItemCount : _draggingIndex - 1 inSection:0]];
            
            [UIView animateWithDuration:0.1 animations:^{
                cell1.transform = CGAffineTransformMakeScale(0.85, 0.85);
                cell2.transform = CGAffineTransformMakeScale(0.85, 0.85);
            }];
            
            NSInteger topIndex = (itemIndex - 1 < 0) ? self.banners.count - 1 : itemIndex - 1;
            NSInteger bottomIndex = itemIndex;
            
            MDBannerModel *model1 = _banners[topIndex];
            [self.topImage sd_setImageWithURL:[NSURL URLWithString:model1.bgImg]];
            MDBannerModel *model2 = _banners[bottomIndex];
            [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model2.bgImg]];
            
            [UIView animateWithDuration:0.5 animations:^{
                [self.leftView setRadius:(mainW * _draggingIndex - curretContentOffset)*2  direction:MDBannerSrollDirectionRight];
                [self.rightView setRadius:0  direction:MDBannerSrollDirectionLeft];
            }];
            
        }else{
            UICollectionViewCell *cell1 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:_draggingIndex inSection:0]];
            UICollectionViewCell *cell2 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_draggingIndex + 1 >= _totalItemCount) ? _totalItemCount/2 : _draggingIndex + 1 inSection:0]];
            
            [UIView animateWithDuration:0.1 animations:^{
                cell1.transform = CGAffineTransformMakeScale(0.85, 0.85);
                cell2.transform = CGAffineTransformMakeScale(0.85, 0.85);
            }];
            
            NSInteger topIndex = (itemIndex + 1 >= self.banners.count) ? 0 : itemIndex + 1;
            NSInteger bottomIndex = itemIndex;
            
            MDBannerModel *model1 = _banners[topIndex];
            [self.topImage sd_setImageWithURL:[NSURL URLWithString:model1.bgImg]];
            MDBannerModel *model2 = _banners[bottomIndex];
            [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model2.bgImg]];
            
            [UIView animateWithDuration:0.5 animations:^{
                [self.leftView setRadius:0 direction:MDBannerSrollDirectionRight];
                [self.rightView setRadius:(mainW * _draggingIndex - curretContentOffset)*2   direction:MDBannerSrollDirectionLeft];
            }];
        }
    }else{
        NSInteger itemIndex = [self pageControlIndexWithCurrentCellIndex:_autoScrollIndex];
        if (_lastContentOffset > curretContentOffset){
           
            //上一个cell
            UICollectionViewCell *cell = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex - 1 < 0) ? _totalItemCount : _autoScrollIndex inSection:0]];
            
            [UIView animateWithDuration:0.1 animations:^{
                cell.transform = CGAffineTransformMakeScale(0.85, 0.85);
            }];
            
            NSInteger topIndex = (itemIndex - 1 < 0) ? self.banners.count - 1 : itemIndex - 1;
            NSInteger bottomIndex = itemIndex;
            
            MDBannerModel *model1 = _banners[topIndex];
            [self.topImage sd_setImageWithURL:[NSURL URLWithString:model1.bgImg]];
            MDBannerModel *model2 = _banners[bottomIndex];
            [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model2.bgImg]];
            
            [UIView animateWithDuration:0.5 animations:^{
                [self.leftView setRadius:(curretContentOffset - mainW * _autoScrollIndex)*2  direction:MDBannerSrollDirectionRight];
                [self.rightView setRadius:0  direction:MDBannerSrollDirectionLeft];
            }];
            
        }else{
            UICollectionViewCell *cell2 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex + 1 >= _totalItemCount) ? _totalItemCount/2 : _autoScrollIndex + 1 inSection:0]];
            
            [UIView animateWithDuration:0.1 animations:^{
                cell2.transform = CGAffineTransformMakeScale(0.85, 0.85);
            }];
            
            NSInteger topIndex = (itemIndex + 1 >= self.banners.count) ? 0 : itemIndex + 1;
            NSInteger bottomIndex = itemIndex;
            
            MDBannerModel *model1 = _banners[topIndex];
            [self.topImage sd_setImageWithURL:[NSURL URLWithString:model1.bgImg]];
            MDBannerModel *model2 = _banners[bottomIndex];
            [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model2.bgImg]];
            
            [UIView animateWithDuration:0.5 animations:^{
                [self.leftView setRadius:0 direction:MDBannerSrollDirectionRight];
                [self.rightView setRadius:(curretContentOffset - mainW * _autoScrollIndex)*2   direction:MDBannerSrollDirectionLeft];
            }];
        }
        _lastContentOffset = curretContentOffset;
    }
}

-(void)scrollViewWillBeginDragging:(UIScrollView *)scrollView{
    self.draggingIndex = [self currentIndex];
    if (_autoScroll){
        [self invalidateTimer];
    }
}


-(void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate{
    if (_autoScroll){
        [self startTimer];
    }
}

-(void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView{
    _lastContentOffset = scrollView.contentOffset.x;
    
    _autoScrollIndex = [self currentIndex];
    
    
    UICollectionViewCell *cell1 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:_autoScrollIndex inSection:0]];
    UICollectionViewCell *cell2 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex + 1 >= _totalItemCount) ? _totalItemCount/2 : _autoScrollIndex + 1 inSection:0]];
    UICollectionViewCell *cell3 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex - 1 < 0) ? _totalItemCount : _autoScrollIndex - 1 inSection:0]];
    
    [UIView animateWithDuration:0.2 animations:^{
        cell1.transform = CGAffineTransformMakeScale(1, 1);
        cell2.transform = CGAffineTransformMakeScale(1, 1);
        cell3.transform = CGAffineTransformMakeScale(1, 1);
    }];
}


-(void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView{
    _autoScrollIndex = [self currentIndex];
    
    UICollectionViewCell *cell1 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:_autoScrollIndex inSection:0]];
    UICollectionViewCell *cell2 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex + 1 >= _totalItemCount) ? _totalItemCount/2 : _autoScrollIndex + 1 inSection:0]];
    UICollectionViewCell *cell3 = [_banner cellForItemAtIndexPath:[NSIndexPath indexPathForItem:(_autoScrollIndex - 1 < 0) ? _totalItemCount : _autoScrollIndex - 1 inSection:0]];
    
    [UIView animateWithDuration:0.2 animations:^{
        cell1.transform = CGAffineTransformMakeScale(1, 1);
        cell2.transform = CGAffineTransformMakeScale(1, 1);
        cell3.transform = CGAffineTransformMakeScale(1, 1);
    }];
}


#pragma Get
-(CGFloat)naxH{
    if (mainH == 812){
        return 88;
    }
    return 64;
}


-(UIImageView *)topImage{
    if (_topImage == nil){
        _topImage = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height - 30)];
    }
    return _topImage;
}

-(UIImageView *)bottomImage{
    if (_bottomImage == nil){
        _bottomImage = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height - 30)];
    }
    return _bottomImage;
}

-(UICollectionViewFlowLayout *)flowLayout{
    if (_flowLayout == nil){
        _flowLayout = [[UICollectionViewFlowLayout alloc] init];
        _flowLayout.itemSize = CGSizeMake(self.width, self.height - self.naxH - 10);
        _flowLayout.minimumLineSpacing = 0;
        _flowLayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    }
    return _flowLayout;
}

-(UICollectionView *)banner{
    if (_banner == nil){
        _banner = [[UICollectionView alloc] initWithFrame:CGRectMake(0, self.naxH + 10, self.width, self.height - self.naxH - 10) collectionViewLayout:self.flowLayout];
        _banner.backgroundColor = [UIColor clearColor];
        _banner.pagingEnabled = YES;
        _banner.showsHorizontalScrollIndicator = NO;
        _banner.delegate = self;
        _banner.dataSource = self;
    }
    return _banner;
}

-(MDMaskView *)leftView{
    if (_leftView == nil){
        _leftView = [[MDMaskView alloc] initWithFrame:CGRectMake(-10, 0, self.width + 10, self.height - 30)];
    }
    return _leftView;
}

-(MDMaskView *)rightView{
    if (_rightView == nil){
        _rightView = [[MDMaskView alloc] initWithFrame:CGRectMake(0, 0, self.width + 10, self.height - 30)];
    }
    return _rightView;
}

-(UIView *)bannerMaskView{
    if (_bannerMaskView == nil){
        _bannerMaskView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height - 30)];
    }
    return _bannerMaskView;
}


-(UIPageControl *)pageControl{
    if (_pageControl == nil){
        _pageControl = [[UIPageControl alloc] init];
        _pageControl.center = CGPointMake(self.center.x, self.y + self.height - 20);
        [_pageControl sizeToFit];
    }
    return _pageControl;
}



#pragma Set
-(void)setBanners:(NSArray *)banners{
    if (banners.count == 0){return;}
    self.totalItemCount = (self.infiniteLoop) ? banners.count * 200 : banners.count;
    
    _banners = banners;
    [self.banner reloadData];
    
    [self.banner scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:_totalItemCount/2 inSection:0] atScrollPosition:UICollectionViewScrollPositionNone animated:NO];
    
    NSInteger index = [self pageControlIndexWithCurrentCellIndex:[self currentIndex]];
    
    self.pageControl.numberOfPages = banners.count;
    self.pageControl.currentPage = index;
    
    if (index + 1 > banners.count - 1){
        MDBannerModel *model = banners[0];
        [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model.bgImg]];
    }else{
        MDBannerModel *model = banners[index + 1];
        [self.bottomImage sd_setImageWithURL:[NSURL URLWithString:model.bgImg]];
    }

    MDBannerModel *model = banners[index];
    [self.topImage sd_setImageWithURL:[NSURL URLWithString:model.bgImg]];

    if (_autoScroll){
        [self startTimer];
    }
}

-(void)setPageSelectImage:(UIImage *)pageSelectImage{
    if (pageSelectImage) {
        [self.pageControl setValue:pageSelectImage forKeyPath:@"_currentPageImage"];
    }else{
        NSLog(@"您设置的pageSelectImage图片是空的");
    }
}

-(void)setPageUnselectImage:(UIImage *)pageUnselectImage{
    if (pageUnselectImage){
        [self.pageControl setValue:pageUnselectImage forKeyPath:@"_pageImage"];
    }else{
        NSLog(@"您设置的pageUnselectImage图片是空的");
    }
}



@end

File: MDShockBannerView/MDShockBannerView.swift



//
//  MDShockBannerView.swift
//  MDShockAnimate
//
//  Created by Alan on 2018/4/17.
//  Copyright © 2018年 MD. All rights reserved.
//

import UIKit
import SDWebImage

let mainW = UIScreen.main.bounds.width
let mainH = UIScreen.main.bounds.height

var navH:CGFloat = 64

let cellID = "MDBannerCollectionViewCell"

class MDShockBannerView: UIView {

    lazy var topImage:UIImageView = {
        let topImage = UIImageView.init(frame: CGRect(x: 0, y: 0, width: self.width, height: self.height - 30))
        topImage.contentMode = .scaleAspectFill
        return topImage
    }()
    
    lazy var bottomImage:UIImageView = {
        let bottomImage = UIImageView.init(frame: CGRect(x: 0, y: 0, width: self.width, height: self.height - 30))
        bottomImage.contentMode = .scaleAspectFill
        return bottomImage
    }()
    
    
    lazy var flowLayout:UICollectionViewFlowLayout = {
        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.itemSize = CGSize(width: self.width, height: self.height - navH - 10)
        flowLayout.minimumLineSpacing = 0
        flowLayout.scrollDirection = .horizontal
        return flowLayout
    }()
    
    lazy var banner:UICollectionView = {
        let banner = UICollectionView.init(frame: CGRect(x:0, y: navH + 10, width: self.width, height: self.height - navH - 10), collectionViewLayout: flowLayout)
        banner.backgroundColor = .clear
        banner.isPagingEnabled = true
        banner.showsHorizontalScrollIndicator = false
        banner.delegate = self
        banner.dataSource = self
        self.addSubview(banner)
        self.bringSubview(toFront: banner)
        return banner
    }()
    
    lazy var pageControl:UIPageControl = {
        let pageControl = UIPageControl()
        pageControl.center = CGPoint(x: self.center.x, y: self.y + self.height - 20)
        pageControl.sizeToFit()
        self.addSubview(pageControl)
        return pageControl
    }()
    
    lazy var leftView:MDMaskView = {
        let leftView = MDMaskView(frame:CGRect(x: -10, y: 0, width: self.width + 10, height: self.height - 30))
        return leftView
    }()
    
    lazy var rightView:MDMaskView = {
        let rightView = MDMaskView(frame:CGRect(x: 0, y: 0, width: self.width + 10, height: self.height - 30))
        return rightView
    }()
    
    lazy var bannerMaskView:UIView = {
        let bannerMaskView = UIView(frame:CGRect(x: 0, y: 0, width: self.width, height: self.height))
        return bannerMaskView
    }()
    
    
    

    var banners:[MDBannerModel]!
    
    var infiniteLoop:Bool = true
    
    var autoScroll:Bool = true
    
    private var timers:Timer?
    
    private var lastContentOffset:CGFloat = 0

    private var  totalItemCount:Int = 0
    
    private var draggingIndex:Int = 0 // 拖动时的索引
    
    private var autoScrollIndex:Int = 0 // 自动滚动时的索引
    

    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    
    func setUI(){
        self.addSubview(self.bottomImage)
        self.addSubview(self.topImage)
        
        self.topImage.mask = bannerMaskView
        
        bannerMaskView.addSubview(rightView)
        bannerMaskView.addSubview(leftView)

        self.banner.register(UINib.init(nibName: "MDBannerCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: cellID)
    }
    
    
    func setBanner(banners:[MDBannerModel]){
        if banners.count > 1{
            self.totalItemCount = self.infiniteLoop ? banners.count * 100 : banners.count
            self.banners = banners
            self.banner.reloadData()
            self.banner.scrollToItem(at: IndexPath.init(row: totalItemCount/2, section: 0), at: [], animated: false)
            
            let index = self.pageControlIndexWithCurrentCellIndex(index: self.currentIndex())
            
            self.pageControl.numberOfPages = banners.count
            self.pageControl.currentPage = index
            self.setPageControlImage()
            
            if index + 1 > banners.count - 1{
                self.bottomImage.sd_setImage(with: URL.init(string: banners[0].bgImg!), completed: nil)
            }else{
                self.bottomImage.sd_setImage(with: URL.init(string: banners[index + index + 1].bgImg!), completed: nil)
            }
            
            self.topImage.sd_setImage(with: URL.init(string: banners[index].bgImg!), completed: nil)
            
            
            self.startTimer()
        }
        
    }
    
    func setPageControlImage(){
        // 初始化一个属性列表数组
//        var ivarName_pageControl: [String] = []
//        var count: uint = 0
//        // 获取属性列表
//        let list = class_copyIvarList(UIPageControl.classForCoder(), &count)
//
//        for index in 0 ... count-1 {
//            // 获取属性名称，ivar_getTypeEncoding 可获取属性类型
//            let ivarName = ivar_getName( list![ Int(index) ] )
//            let name = String.init(cString: ivarName!)
//            print(name)
//            ivarName_pageControl.append(name)
//        }
//        // 判断是否包含这两个属性
//        if ivarName_pageControl.contains("_pageImage") && ivarName_pageControl.contains("_currentPageImage")
//        {
//            pageControl.setValue(UIImage.init(named: "banner_unselect"), forKey: "_pageImage")
//            pageControl.setValue(UIImage.init(named: "banner_select"), forKey: "_currentPageImage")
//            pageControl.layoutIfNeeded()
//        }
    }
    
    func startTimer(){
        self.invalidateTimer()
        self.autoScrollIndex  = self.currentIndex()
        timers = Timer.init(timeInterval: 3, target: self, selector: #selector(automaticScroll), userInfo: nil, repeats: true)
        RunLoop.main.add(timers!, forMode: .commonModes)
    }
    
    func invalidateTimer(){
        if timers != nil{
            timers?.invalidate()
            timers = nil
        }
    }
    
    @objc func automaticScroll(){
        if totalItemCount == 0{return}
        let currentIndex = self.currentIndex()
        let targetIndex = currentIndex + 1
        self.scrollToIndex(targetIndex: targetIndex)
    }
    
    
    func scrollToIndex(targetIndex:Int){
        if targetIndex >= totalItemCount{
            banner.scrollToItem(at: IndexPath.init(row: totalItemCount / 2, section: 0), at: [], animated: false)
            return
        }
        banner.scrollToItem(at: IndexPath.init(row: targetIndex, section: 0), at: [], animated: true)
    }
    
    func pageControlIndexWithCurrentCellIndex(index:Int) ->Int{
        return index % self.banners.count
    }
}



extension MDShockBannerView:UICollectionViewDelegate,UICollectionViewDataSource{
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return totalItemCount
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellID, for: indexPath) as!MDBannerCollectionViewCell
        let itemIndex =  self.pageControlIndexWithCurrentCellIndex(index: indexPath.item)
        let banner = self.banners[itemIndex]
        cell.i_image.sd_setImage(with: URL.init(string: banner.img!), completed: nil)
        return cell
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let curretContentOffset = scrollView.contentOffset.x
        self.pageControl.currentPage = self.pageControlIndexWithCurrentCellIndex(index: self.currentIndex())
        if scrollView.isDragging || scrollView.isDecelerating{
            let itemIndex = self.pageControlIndexWithCurrentCellIndex(index: draggingIndex)
        
            if lastContentOffset > curretContentOffset{
                
                let cell1 = banner.cellForItem(at: IndexPath.init(row: self.draggingIndex, section: 0))
                let cell2 = banner.cellForItem(at: IndexPath.init(row: self.draggingIndex - 1, section: 0))
                
                UIView.animate(withDuration: 0.1) {
                    cell1?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                    cell2?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                }
                
                
                let topIndex = itemIndex - 1 < 0 ? self.banners.count - 1 : itemIndex - 1
                let bottomIndex = itemIndex
                self.topImage.sd_setImage(with: URL.init(string: banners[topIndex].bgImg!), completed: nil)
                self.bottomImage.sd_setImage(with: URL.init(string: banners[bottomIndex].bgImg!), completed: nil)
                UIView.animate(withDuration: 0.5, animations: {
                    self.leftView.setRadius(radius: (mainW * CGFloat(self.draggingIndex) - curretContentOffset) * 2, direction: .right)
                    self.rightView.setRadius(radius: 0, direction: .left)
                })
                
            }else{
                
                let cell1 = banner.cellForItem(at: IndexPath.init(row: self.draggingIndex, section: 0))
                let cell2 = banner.cellForItem(at: IndexPath.init(row: self.draggingIndex + 1, section: 0))
                
                UIView.animate(withDuration: 0.1) {
                    cell1?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                    cell2?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                }
                
                let topIndex = itemIndex + 1 > self.banners.count - 1 ? 0 : itemIndex + 1
                let bottomIndex = itemIndex
                self.topImage.sd_setImage(with: URL.init(string: banners[topIndex].bgImg!), completed: nil)
                self.bottomImage.sd_setImage(with: URL.init(string: banners[bottomIndex].bgImg!), completed: nil)
                UIView.animate(withDuration: 0.5, animations: {
                    self.leftView.setRadius(radius: 0, direction: .right)
                    self.rightView.setRadius(radius: (mainW * CGFloat(self.draggingIndex) - curretContentOffset) * 2, direction: .left)
                })
            }
        }else{
            let itemIndex = self.pageControlIndexWithCurrentCellIndex(index: autoScrollIndex)
            if lastContentOffset > curretContentOffset{
                
                let cell1 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex, section: 0))
                let cell2 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex - 1, section: 0))
                
                UIView.animate(withDuration: 0.1) {
                    cell1?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                    cell2?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                }
                
                let topIndex = itemIndex - 1 < 0 ? self.banners.count - 1 : itemIndex - 1
                let bottomIndex = itemIndex
                self.topImage.sd_setImage(with: URL.init(string: banners[topIndex].bgImg!), completed: nil)
                self.bottomImage.sd_setImage(with: URL.init(string: banners[bottomIndex].bgImg!), completed: nil)
                UIView.animate(withDuration: 0.5, animations: {
                    self.leftView.setRadius(radius: (curretContentOffset - mainW * CGFloat(itemIndex)) * 2, direction: .right)
                    self.rightView.setRadius(radius: 0, direction: .left)
                })
                
            }else{
                
                let cell1 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex, section: 0))
                let cell2 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex + 1, section: 0))
                
                UIView.animate(withDuration: 0.1) {
                    cell1?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                    cell2?.transform = CGAffineTransform(scaleX: 0.85, y: 0.85)
                }
                
                let topIndex = (itemIndex + 1) > self.banners.count - 1 ? 0 : itemIndex + 1
                let bottomIndex = itemIndex
                self.topImage.sd_setImage(with: URL.init(string: banners[topIndex].bgImg!), completed: nil)
                self.bottomImage.sd_setImage(with: URL.init(string: banners[bottomIndex].bgImg!), completed: nil)
                UIView.animate(withDuration: 0.5, animations: {
                    self.leftView.setRadius(radius: 0, direction: .right)
                    self.rightView.setRadius(radius: (curretContentOffset - mainW * CGFloat(self.autoScrollIndex)) * 2, direction: .left)
                })
            }
             lastContentOffset = scrollView.contentOffset.x
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
       
    }
    
    func collectionView(_ collectionView: UICollectionView, didDeselectItemAt indexPath: IndexPath) {

    }
    

    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        self.draggingIndex = self.currentIndex()
        if self.autoScroll{
            self.invalidateTimer()
        }
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if self.autoScroll{
            self.startTimer()
        }
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        //拖动结束后
        lastContentOffset = scrollView.contentOffset.x
        self.autoScrollIndex = self.currentIndex()
    
        let cell1 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex - 1 , section: 0))
        let cell2 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex , section: 0))
        let cell3 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex  + 1, section: 0))
        
        UIView.animate(withDuration: 0.1) {
            cell1?.transform = CGAffineTransform(scaleX: 1, y: 1)
            cell2?.transform = CGAffineTransform(scaleX: 1, y: 1)
            cell3?.transform = CGAffineTransform(scaleX: 1, y: 1)
        }
    }
    
    
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        self.autoScrollIndex = self.currentIndex()
        
        let cell1 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex - 1 , section: 0))
        
        let cell2 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex , section: 0))
        let cell3 = banner.cellForItem(at: IndexPath.init(row: self.autoScrollIndex  + 1, section: 0))
        
        UIView.animate(withDuration: 0.1) {
            cell1?.transform = CGAffineTransform(scaleX: 1, y: 1)
            cell2?.transform = CGAffineTransform(scaleX: 1, y: 1)
            cell3?.transform = CGAffineTransform(scaleX: 1, y: 1)
        }
    }
    
    
    
    
    func currentIndex() ->Int{
        if banner.width == 0 || banner.height == 0{
            return 0
        }
        var index:Int = 0
        index = Int((banner.contentOffset.x + flowLayout.itemSize.width / 2) / flowLayout.itemSize.width)
        return max(0, index)
    }
    
}
